### ISSUE TITLE
[Manuscript] General Revision and comments (openjournals/joss-reviews#4822)

### ISSUE THREAD

**Issue by rspavel** [lanl/GLUE#14]
Recreation of #6 as issue on behalf of @govarguz 

----

Hello everyone,

this a very creative framework for handling multiscale communication. We had also a similar multiscale challenge by sequentially going back and forth between AA-MD (also CG-MD) simulations and Poisson Boltzmann [HERE](https://github.com/pyF4all/Notebooks_phase1). However, we did not manage the automated communication included in GLUE, if I understand it correctly?

I will start now with general comments:

1. There is a conceptual difference between sequential and concurrent multiscale simulations, looking at figure 1 I would rather say GLUE seems to be sequential, at least until a 'specific model' has been trained, did I get this correctly?  Is there a concurrent capability of GLUE?  In any case please clarify shortly in page 1, section: Summary. 
2. In the same lines, solving the load imbalance issue is indeed key for large scale simulations, nonetheless, another important point is the initial spatial decomposition of the simulation boxes, for this, it was shown by predictive domain decomposition that heterogeneous systems can be also initially optimized, which is not dynamic load balancing but still a good starting point, specially for concurrent multiscale is worth mentioning https://doi.org/10.1103/PhysRevE.96.053311
3. Does GLUE include a parser for different scale formats? If not, how is this handled?

This so far, thanks in advance for your point-to-point answers!

Part of https://github.com/openjournals/joss-reviews/issues/4822

**Comment by rspavel** [comment]
Yeah. Sorry was focusing on the issues.

### Regarding concurrency: 

This is a somewhat complicated topic and speaks to the different parts of the overall system as depicted in Figure 1. The parallelism of the **Coarse Grain Solver** is fully dependent on the solver itself but the `GLUECode_Library` provides MPI interfaces  to support distributed applications and all couplings thus far (barring hand-written tests for CI purposes) were with distributed coarse grain simulations. There was work on providing a thread-safe interface for on-node parallelism but this was given low priority because...

A single `GLUECode_Service` itself is primarily a series of sequential operations that occur in a persistent service and is built around monitoring a task queue. Said task queue allows for parallelism via the use of HPC **Job Scheduler**s, but the actual computation/generation of ML models as well as processing of requests and results is sequential;. That said, multiple `GLUECode_Service` instances can be run in parallel and this has been done as a way to lessen congestion for particularly large simulation.

And the **Fine Grain Solver** is once again dependent on the solver. That said, we have done work with both gpu-enabled LAMMPS as well as MPI enabled LAMMPS and the `SlurmScheduler` and `FluxScheduler` parts of our json schema (https://github.com/lanl/GLUE/blob/manuscript/docs/inputSchema.json) are specifically set up to provide these configurations.

But, generally speaking, the overall `GLUECode` has a high degree of concurrency and tends to resemble fork-join parallelism/MapReduce in practice even if it consists of sequential tasks/stages.


### Regarding load imbalance:

Fully agree. We mostly thought back to the traditional approaches used by asynchronous task based runtimes that, generally speaking, rely on some form of a task queue and then various work stealing algorithms. But, due to the granularity of our tasks we chose to cheat a bit and offload that to HPC job schedulers like Slurm and Flux. The idea being that by putting this work into a job queue we can use proven tools to schedule those in an efficient manner.

In terms of domain aware scheduling and optimizations: I would need to check exactly how in-depth the work on the ICF side went into it, but there was some work put in to preprocessing the requests prior to calling functions/subroutines like `bgk_req_batch_subr_f`. But part of the goal from a CS and ML perspective was to stress test the overall `GLUECode` interface as well as evaluate the benefits of active learning based models.

And thanks. We'll definitely take a look at that paper.


### Multiple formats/parsing tools:

Generally speaking, we only provide the interfaces required for the coupling we are working on as one of the primary goals of this project was to provide a minimally invasive tool. Domain experts tend to be the most knowledgeable in terms of what they need from each level of the simulation so we largely use co-design to define interfaces (https://github.com/lanl/GLUE/blob/main/GLUECode_Library/src/include/alInterface.h being the C-side versions) as well as using control variables in the overall system configuration to utilize the appropriate tools (e.g. https://github.com/lanl/GLUE/blob/main/GLUECode_Service/processBGKResult.py) to generate and post-process results.

Some thought was put into making even a general purpose interface to LAMMPS (or just using the existing python bindings) but due to a mix of wanting to be able to run these jobs remotely (currently across jobs in an HPC cluster) it was deemed to be easier to just work with domain experts to generate scripts as needed for the various jobs along the lines of https://github.com/lanl/GLUE/blob/main/GLUECode_Service/ICF_Utils.py#L1001.

As new couplings are added the existing tools will be generalized as needed. But we would much rather domain experts take advantage of their knowledge of tools like LAMMPS rather than rely on us providing all of the interfaces they need.

**Comment by apachalieva** [comment]
We added information about concurrency, load imbalance, and formatting/parsing to the manuscript. With this, we close this issue.
